spec.bs

```diff
@@ -8159,43 +8324,151 @@ To <dfn>batch or fetch trusted bidding signals</dfn> given a [=trusted bidding s
+    :: "experiment_group_id" → |experimentGroupId|
+    :: |slotSizeParams|[0] → |slotSizeParams|[1]
+  1. Otherwise:
+    1. [=list/Append=] |group|'s [=interest group/name=] into |compressionGroupMap|[|compressionGroupId|]
+      [|partitionId|]["interest_group_names"].
+    1. [=list/Append=] |group|'s [=interest group/trusted bidding signals keys=] into
+      |compressionGroupMap|[|compressionGroupId|][|partitionId|]["bidding_keys"].
+1. [=map/For each=] |id| → |group| in |compressionGroupMap|:
+  1. Let |compressionGroup| be an empty [=bidding compression group=].
+  1. Set |compressionGroup|'s [=bidding compression group/compression group id=] to |id|.
+  1. Set |compressionGroup|'s [=bidding compression group/partition=] to an empty [=list=].
+  1. [=list/For each=] |partition| in |group|'s [=map/values=]:
+    1. [=list/Append=] |partition| to |compressionGroup|'s [=bidding compression group/partition=].
+  1. [=list/Append=] |compressionGroup| to |compressionGroups|.
+1. Let |metadata| be an empty [=map=], whose [=map/keys=] and [=map/values=] are [=strings=].
+1. Let |hostname| be the result of [=string/UTF-8 percent-encoding=] the
+  [=serialization of an origin|serialized=] |topLevelOrigin| using [=component percent-encode set=].
+1. [=map/Set=] |metadata|["hostname"] to |hostname|.
+1. Let |keyInfo| be the result of [=looking up the server encryption key=] with |owner| and
+  |coordinator|.
+1. If |keyInfo| is failure, then return « null, null, null ».
+1. Let (|requestBlob|, |context|) be the result of generating request with |keyInfo|, |metadata| and
+  |compressionGroups|. The generation method may follow that described in 
+  [Section 2.2.4 of the Protected Audience Key Value Services](https://privacysandbox.github.io/draft-ietf-protected-audience-key-value-service/draft-ietf-protected-audience-key-value-services.html#name-generating-a-request).
+1. Return « |requestBlob|, |interestGroupIdMap|, |context| ».
+
+</div>
+
+<div algorithm>
+To <dfn>fetch the trusted key value signals batch</dfn> given a
+[=trusted bidding signals batcher=] |trustedBiddingSignalsBatcher|, a [=URL=] |signalsUrl|,
+an [=origin=] |scriptOrigin|, an {{unsigned short}}-or-null |experimentGroupId|, an [=origin=]
+|topLevelOrigin|, a [=string=] |slotSizeQueryParam|, and a [=policy container=] |policyContainer|:
+
+  1. If |signalsUrl| is null, return.
+  1. [=map/For each=] (|joiningOrigin|, |owner|) → |interestGroups| of [=trusted bidding signals
+    batcher/key value interest groups=]:
+    1. Let « |requestBody|, |interestGroupIdMap|, |context| » be the result of [=building trusted
+      key value bidding signals request body=] with |interestGroups|, |experimentGroupId|,
+      |topLevelOrigin|, |slotSizeQueryParam|, |joiningOrigin| and |owner|.
+    1. Let « |partialTrustedBiddingSignals|, |partialPerInterestGroupData|, |dataVersion| » be the
+      result of [=fetching trusted key value signals=] with |signalsUrl|, |requestBody|, |context|,

```

Comment:
same comment. Now `|requestBody|, |renderUrlIdMap|, |context|` can all be null, so you'd either update `[=fetching trusted key value signals=]`'s definition to allow these parameters to be null, or just check whether |requestBody| is null here, and only call the algorithm when it's not null.

Reply:
See the comment https://github.com/WICG/turtledove/pull/1340/files#r1888945731

Reply:
similar to that comment, it would be better to only fetch when body is not null:
```
1. Let « |requestBody| ... be ....[=building ...
1. If |requestBody| is null, then let |partialTrustedBiddingSignals| be null.
1. Otherwise, let « |partialTrustedBiddingSignals| .... [=fetching ...
1. ...
```